### PR TITLE
add breaking change notice for 'az deployment'

### PR DESCRIPTION
**Pull request by Tiano2017**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by Juliehzl** [comment]
@Tiano2017 You can run `azdev style resource` locally to find style errors in your computer and save time for CI check.

**Comment by Tiano2017** [comment]
addressed style issue, but seems there's something wrong with the implementation. right now the breaking change notice is printed even when I do "az group". Need to get a better solution.

**Comment by Juliehzl** [comment]
@Tiano2017 When I test your code locally, it looks you also add warning message for all CLI commands. Can you confirm if the warning message only show up when you run `az deployment` commands?  You can try `az -h` in your virtual environment to test it. 
In addition, if you want to remove `az group deployment` commands as what we have discussed before, you also need to add deprecate info for it.

**Comment by Tiano2017** [comment]
@Juliehzl I want it to only show for 'az deployment', but right now it seems it shows for all commands. I'm doing "logger.warning" in "def command_group". Do you have a pointer of where I should do the "logger.warning"?

As for 'az group deployment', we'll need to add the notice later in a separate PR when update the functionality of 'az deployment'.

**Comment by Juliehzl** [comment]
> @Juliehzl I want it to only show for 'az deployment', but right now it seems it shows for all commands. I'm doing "logger.warning" in "def command_group". Do you have a pointer of where I should do the "logger.warning"?
> 
> As for 'az group deployment', we'll need to add the notice later in a separate PR when update the functionality of 'az deployment'.

Hi @Tiano2017, I think maybe you cannot add logging.warning in command_group, because command_group will be run for every command. Typically we add logger.warning in custom.py which will be called only when corresponding command is running. 

In your scenario, you want to work for one command group. Why not try to add  [deprecate_info](https://github.com/Azure/azure-cli/blob/dev/doc/authoring_command_modules/authoring_commands.md#deprecating-commands-and-arguments) kwargs for your command group here?

**Comment by Tiano2017** [comment]
> > @Juliehzl I want it to only show for 'az deployment', but right now it seems it shows for all commands. I'm doing "logger.warning" in "def command_group". Do you have a pointer of where I should do the "logger.warning"?
> > As for 'az group deployment', we'll need to add the notice later in a separate PR when update the functionality of 'az deployment'.
> 
> Hi @Tiano2017, I think maybe you cannot add logging.warning in command_group, because command_group will be run for every command. Typically we add logger.warning in custom.py which will be called only when corresponding command is running.
> 
> In your scenario, you want to work for one command group. Why not try to add [deprecate_info](https://github.com/Azure/azure-cli/blob/dev/doc/authoring_command_modules/authoring_commands.md#deprecating-commands-and-arguments) kwargs for your command group here?

Thanks @Juliehzl. I tried to use "deprecate_info", but it always gives a message like "This command group has been deprecated and will be removed in a future release. Use 'az xxx' instead." The message is not accurate as I don't want to deprecate the command. I just want to do a breaking change. Is it possible to customize the message given by "deprecate_info"?

**Comment by Juliehzl** [comment]
Hi @Tiano2017, can you add your [breaking change] message through long-summary part in [_help](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/resource/_help.py#L209)? In addition, you can add logger.warning for each command under "deployment" group in custom.py file.

**Comment by Tiano2017** [comment]
> Hi @Tiano2017, can you add your [breaking change] message through long-summary part in [_help](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/resource/_help.py#L209)? In addition, you can add logger.warning for each command under "deployment" group in custom.py file.

sure. will try it out.

**Comment by Tiano2017** [comment]
> Hi @Tiano2017, can you add your [breaking change] message through long-summary part in [_help](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/resource/_help.py#L209)? In addition, you can add logger.warning for each command under "deployment" group in custom.py file.

updated.

**Comment by Tiano2017** [comment]
@Juliehzl could you take another look at the PR? BTW,  when is the next CLI release scheduled?

File: src/azure-cli/azure/cli/command_modules/resource/commands.py [hunk 1]
```diff
@@ -232,16 +232,16 @@ def load_command_table(self, _):
...
-        g.command('list', 'list_at_subscription_scope', table_transformer=transform_deployments_list)
-        g.show_command('show', 'get_at_subscription_scope')
-        g.command('delete', 'delete_at_subscription_scope', supports_no_wait=True)
+        g.custom_command('list', 'list_deployments_at_subscription_scope', table_transformer=transform_deployments_list)
+        g.custom_show_command('show', 'get_deployment_at_subscription_scope')
+        g.custom_command('delete', 'delete_deployment_at_subscription_scope', supports_no_wait=True)
         g.custom_command('validate', 'validate_arm_template_at_subscription_scope', table_transformer=deployment_validate_table_format, exception_handler=handle_template_based_exception)
         g.custom_command('create', 'deploy_arm_template_at_subscription_scope', supports_no_wait=True, validator=process_deployment_create_namespace, exception_handler=handle_template_based_exception)
         g.custom_command('export', 'export_subscription_deployment_template')
         g.wait_command('wait', getter_name='get_at_subscription_scope')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/resource/commands.py]: What about wait command? Do you want to involve the warning message in wait command?

File: src/azure-cli/azure/cli/command_modules/resource/_help.py [hunk 2]
```diff
@@ -209,6 +209,34 @@
 helps['deployment'] = """
 type: group
 short-summary: Manage Azure Resource Manager deployments at subscription scope.
+long-summary: >
+    Upcoming breaking change: a new parameter "scope-type" will be introduced to commands in this group and will be mandatory.
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/resource/_help.py]: Can you use a special sign to make the breaking change more conspicuous? e.g. [UPCOMING BREAKING CHANGE] or something else.

File: src/azure-cli/azure/cli/command_modules/resource/commands.py (hunk 1 above)
Comment (Tiano2017) [reply src/azure-cli/azure/cli/command_modules/resource/commands.py]: sure. the only trick there is I need to make sure the warning is only logged once. I've updated it. let me know if it looks fine.

File: src/azure-cli/azure/cli/command_modules/resource/_help.py (hunk 2 above)
Comment (Tiano2017) [reply src/azure-cli/azure/cli/command_modules/resource/_help.py]: sounds good. I made it like [**Upcoming breaking change**]

File: src/azure-cli/azure/cli/command_modules/resource/commands.py (hunk 1 above)
Comment (Juliehzl) [reply src/azure-cli/azure/cli/command_modules/resource/commands.py]: According to your code, your warning message only be shown when the corresponding custom method is called. You can test it in your test environment and if you think it is good for you to show breaking change in this way, I can help you merge it now. NO other concern for me. 😊

Comment (Tiano2017) [reply src/azure-cli/azure/cli/command_modules/resource/commands.py]: yeah I tested it and I think it worked properly now. If you look at custom.py line 753, I have flag there to indicate there's it's the first run of the method. I think we can merge it. :)

Comment (Juliehzl) [reply src/azure-cli/azure/cli/command_modules/resource/commands.py]: Sounds Great!

**Pull request merged**